hack/generator/pkg/codegen/storage/conversion_graph.go

```diff
@@ -35,19 +35,54 @@ func (graph *ConversionGraph) LookupTransition(start astmodel.PackageReference)
 	return subgraph.LookupTransition(start)
 }
 
-// FindHubTypeName returns the type name of the hub resource, given the type name of one of the resources that is
-// persisted using that hub type. This is done by following links in the conversion graph until we reach the end
-func (graph *ConversionGraph) FindHubTypeName(name astmodel.TypeName) astmodel.TypeName {
+// FindNext returns the type name of the next closest type on the path to the hub type.
+// Returns the type name and true if the next type is found; an empty name and false if not.
+// If the name passed in is for the hub type for the given resource, no next type will be found.
+// This is used to identify the next type needed for property assignment functions, and is a building block for
+// identification of hub types.
+func (graph *ConversionGraph) FindNext(name astmodel.TypeName, types astmodel.Types) (astmodel.TypeName, bool) {
 	ref := name.PackageReference
 	for {
+		// Find the next package to consider
 		r, ok := graph.LookupTransition(ref)
 		if !ok {
-			break
+			// No next reference
+			return astmodel.TypeName{}, false
 		}
+
+		// Look up to see if the resource exists in this package
+		n := astmodel.MakeTypeName(r, name.Name())
+		if _, ok := types.TryGet(n); ok {
+			// found the next type
+			return n, true
+		}
+
+		// Didn't find it, check the next package
+		// We do this to allow for gaps where a type is dropped and then restored across versions
 		ref = r
 	}
+}
+
+// FindHub returns the type name of the hub resource, given the type name of one of the resources that is
+// persisted using that hub type. This is done by following links in the conversion graph until we either reach the end
+// or we find that a newer version of the type does not exist.
+// Returns the hub type and true if found; an empty name and false if not.
+// If the name passed in is for the hub type for the given resource, no hub type will be found.
```

Comment:
These return values are confusing. If you call `FindHub` with the hub, it should return `(hub, true)`.
If you call it with a non-hub name, it should return `(hub, false)`.

Reply:
I wrote the function to make it easier to consume, but actually I agree with you. Will change.

Reply:
Done.